contracts/dev-utils/contracts/src/OrderValidationUtils.sol

```diff
@@ -0,0 +1,190 @@
+                ? getPartialAmountFloor(
+                    transferableMakerAssetAmount,
+                    order.makerAssetAmount,
+                    takerAssetAmount
+                )
+                : min256(
+                    getPartialAmountFloor(
+                        transferableMakerAssetAmount,
+                        order.makerAssetAmount,
+                        takerAssetAmount
+                    ),
+                    getPartialAmountFloor(
+                        transferableMakerFeeAssetAmount,
+                        makerFee,
+                        takerAssetAmount
+                    )
+                );
+        }
+    
+        // `fillableTakerAssetAmount` is the lower of the order's remaining `takerAssetAmount` and the `purchasableTakerAssetAmount`
+        fillableTakerAssetAmount = min256(
+            safeSub(takerAssetAmount, orderInfo.orderTakerAssetFilledAmount),
+            purchasableTakerAssetAmount
+        );
+
+        return (orderInfo, fillableTakerAssetAmount, isValidSignature);
+    }
+
+    /// @dev Fetches information for all passed in orders and the makers/takers of each order.
+    /// @param orders Array of order specifications.
+    /// @param signatures Proofs that orders have been created by makers.
+    /// @return Arrays of OrderInfo, fillable takerAssetAmounts, and validity of signatures that correspond to each order.
+    function getOrderRelevantStates(LibOrder.Order[] memory orders, bytes[] memory signatures)
+        public
+        view
+        returns (
+            LibOrder.OrderInfo[] memory ordersInfo,
+            uint256[] memory fillableTakerAssetAmounts,
+            bool[] memory isValidSignature
+        )
+    {
+        uint256 length = orders.length;
+        ordersInfo = new LibOrder.OrderInfo[](length);
+        fillableTakerAssetAmounts = new uint256[](length);
+        isValidSignature = new bool[](length);
+
+        for (uint256 i = 0; i != length; i++) {
+            (ordersInfo[i], fillableTakerAssetAmounts[i], isValidSignature[i]) = getOrderRelevantState(
+                orders[i],
+                signatures[i]
+            );
+        }
+
+        return (ordersInfo, fillableTakerAssetAmounts, isValidSignature);
+    }
+
+    /// @dev Gets the address of the AssetProxy that corresponds to the given assetData.
+    /// @param assetData Description of tokens, per the AssetProxy contract specification.
+    /// @return Address of the AssetProxy contract.
+    function getAssetProxyAddress(bytes memory assetData)
+        public
+        view
+        returns (address assetProxyAddress)
+    {
+        if (assetData.equals(ZRX_ASSET_DATA)) {
+            return ERC20_PROXY_ADDRESS;
+        }
+        bytes4 assetProxyId = assetData.readBytes4(0);
+        assetProxyAddress = EXCHANGE.getAssetProxy(assetProxyId);
+        return assetProxyAddress;
+    }
+
+    /// @dev Gets the amount of an asset transferable by the owner.
```

Comment:
Let's mention how multiAsset is handled here. It's important for devs to understand.